mmaction/models/losses/ssn_loss.py

```diff
@@ -0,0 +1,133 @@
+        incomplete_pred = pred[:, positive_per_video:, :].contiguous().view(-1, pred_dim)  # noqa:E501
+        # yapf:enable
+
+        positive_loss = OHEMHingeLoss.apply(
+            positive_pred, gt[:, :positive_per_video].contiguous().view(-1), 1,
+            1.0, positive_per_video)
+        incomplete_loss = OHEMHingeLoss.apply(
+            incomplete_pred, gt[:, positive_per_video:].contiguous().view(-1),
+            -1, ohem_ratio, incomplete_per_video)
+        num_positives = positive_pred.size(0)
+        num_incompletes = int(incomplete_pred.size()[0] * ohem_ratio)
+
+        return ((positive_loss + incomplete_loss) /
+                float(num_positives + num_incompletes))
+
+    def classwise_regression_loss(self, bbox_pred, labels, bbox_targets,
+                                  regression_indexer):
+        """Classwise Regression Loss.
+
+        It will calculate classwise_regression loss given class_reg_pred and
+        targets.
+        """
+        pred = bbox_pred[regression_indexer, :, :]
+        gt = labels[regression_indexer]
+        reg_target = bbox_targets[regression_indexer, :]
+
+        class_idx = gt.data - 1
+        classwise_pred = pred[:, class_idx, :]
+        classwise_reg_pred = torch.cat(
+            (torch.diag(classwise_pred[:, :, 0]).view(
+                -1, 1), torch.diag(classwise_pred[:, :, 1]).view(-1, 1)),
+            dim=1)
+        loss = F.smooth_l1_loss(
+            classwise_reg_pred.view(-1), reg_target.view(-1)) * 2
+        return loss
+
+    def forward(self, activity_score, completeness_score, bbox_pred,
+                proposal_type, labels, bbox_targets, train_cfg):
+
+        self.sampler = train_cfg.ssn.sampler
+        self.loss_weight = train_cfg.ssn.loss_weight
+        losses = dict()
+
+        proposal_type = proposal_type.view(-1)
+        labels = labels.view(-1)
+        activity_indexer = ((proposal_type == 0) +
+                            (proposal_type == 2)).nonzero().squeeze()
+        completeness_indexer = ((proposal_type == 0) +
+                                (proposal_type == 1)).nonzero().squeeze()
+
+        total_ratio = (
+            self.sampler.positive_ratio + self.sampler.background_ratio +
+            self.sampler.incomplete_ratio)
+        positive_per_video = int(self.sampler.num_per_video *
+                                 (self.sampler.positive_ratio / total_ratio))
+        background_per_video = int(
+            self.sampler.num_per_video *
+            (self.sampler.background_ratio / total_ratio))
+        incomplete_per_video = (
+            self.sampler.num_per_video - positive_per_video -
+            background_per_video)
+
+        losses['loss_activity'] = self.activity_loss(activity_score, labels,
+                                                     activity_indexer),
+        losses['loss_activity'] = torch.tensor(
+            device=activity_score.device, requires_grad=True)
```

Comment:
Miss `x` and this operation may be redundant once we early set the `device` and `requires_grad` when calculating the loss.

Reply:
Same to `loss_completeness` and `loss_reg`.